controlplane/pkg/tests/endpoint_service_test.go

```diff
@@ -0,0 +1,250 @@
+	g := NewWithT(t)
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	testEndpoint, nseConn := getTestEndpointAndNSEConnection()
+
+	// Mock NSE client
+	nseClientMock := mock.NewMockNetworkServiceClient(ctrl)
+	nseClientMock.EXPECT().Cleanup().Return(nil)
+	nseClientMock.EXPECT().Request(gomock.Any(), gomock.Any()).DoAndReturn(func(_ context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
+		if request == nil {
+			return nil, errors.Errorf("Request should not be nil")
+		}
+		return nseConn, nil
+	})
+
+	// Mock NSE manager
+	nseManagerMock := mock.NewMockNetworkServiceEndpointManager(ctrl)
+	nseManagerMock.EXPECT().IsLocalEndpoint(gomock.Any()).Return(false)
+	nseManagerMock.EXPECT().CreateNSEClient(gomock.Any(), gomock.Any()).DoAndReturn(
+		func(_ context.Context, endpoint *registry.NSERegistration) (nsm.NetworkServiceClient, error) {
+			if endpoint != testEndpoint {
+				return nil, errors.Errorf("Given endpoint doesn't equal to expected")
+			}
+			return nseClientMock, nil
+		},
+	)
+
+	// Initialize context for the request
+	ctx := common.WithForwarder(context.Background(), testForwarder1)
+	ctx = common.WithLog(ctx, logrus.New())
+	ctx = common.WithModelConnection(ctx, &model.ClientConnection{})
+	ctx = common.WithEndpoint(ctx, testEndpoint)
+
+	// Initialize model
+	testModel := newTestModel()
+	testModel.AddEndpoint(context.Background(),
+		&model.Endpoint{
+			Endpoint:  testEndpoint,
+			Workspace: "workspace",
+		},
+	)
+
+	builder := common.NewLocalRequestBuilder(testModel)
+	service := common.NewEndpointService(nseManagerMock, nil, testModel, builder)
+
+	g.Expect(service).NotTo(BeNil())
+
+	request := &networkservice.NetworkServiceRequest{Connection: getTestConnection()}
+	conn, err := service.Request(ctx, request)
+
+	g.Expect(err).To(BeNil())
+
+	// Check that source connection context is updated and the same as dst context
+	g.Expect(conn.Context).To(Equal(nseConn.Context))
+
+	// Check nse connection parameters are not updated
+	g.Expect(nseConn.GetMechanism().GetParameters()[mechanismCommon.Workspace]).To(Equal(""))
+	g.Expect(nseConn.GetMechanism().GetParameters()[kernel.WorkspaceNSEName]).To(Equal(""))
+}
+
+func TestEndpointServiceCloseConnection(t *testing.T) {
+	g := NewWithT(t)
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	testEndpoint, nseConn := getTestEndpointAndNSEConnection()
+
+	clientConnection := &model.ClientConnection{
+		Endpoint: testEndpoint,
+		Xcon:     crossconnect.NewCrossConnect("id", "payload", nil, nseConn),
+	}
+
+	// Mock NSE client
+	nseClientMock := mock.NewMockNetworkServiceClient(ctrl)
+
+	// Flag to check if Close() was called on client
+	closeCalled := false
+
+	nseClientMock.EXPECT().Close(gomock.Any(), gomock.Any()).DoAndReturn(func(_ context.Context, _ *connection.Connection) error {
+		closeCalled = true
+		return nil
+	})
+
+	// Mock NSE manager
+	nseManagerMock := mock.NewMockNetworkServiceEndpointManager(ctrl)
```

Comment:
Why we can't use for this test already existing stubs?
https://github.com/networkservicemesh/networkservicemesh/blob/1479afb615aca94ae3de155249d91258c21722ab/controlplane/pkg/nsm/nsm_heal_processor_test.go#L388
https://github.com/networkservicemesh/networkservicemesh/blob/1479afb615aca94ae3de155249d91258c21722ab/controlplane/pkg/nsm/nsm_heal_processor_test.go#L399

Reply:
The realizations of `NetworkServiceEndpointManager`'s methods  in `nseManagerMock` and `nseManagerStub` differ, in my tests I need more simple logic for this methods, and some of them are not even called. 
For NSEClient, I need  `Close()` and `Reuquest()` methods, which are not implemented in `nseClientStub`. Of cource, I could add implementation, but I prefer to use mock opportunity. With usage of mocks, we will be able to get rid even of these `nseClientStub` `nseManagerStub` stubs too, and I think it will make code more brief and readable.

Reply:
>ith the usage of mocks, we will be able to get rid even of these nseClientStub nseManagerStub stubs too, 

@a-bastrykina sounds good to me